### PR TITLE
Snapshot for IoT.js

### PR DESCRIPTION
**Pull request by chunseoklee**
1. This commit is based on e3532decdb0fbb8b042e2509698e9b466cf41071 by
Ruben Ayrapetyan.
2. cflags(-fno-short-enums) for arm-nuttx is added.
3. Now, snapshot is supported on default.

IoT.js-DCO-1.0-Signed-off-by: Chunseok Lee chunseok.lee@samsung.com

### CODE REVIEW

**Comment by seanshpark** [comment]
There are segment fault with Raspberry Pi 2 and Ubuntu 32bit.

**Comment by chunseoklee** [comment]
@seanshpark rpi2 issue is fixed but, i cannot test x86

**Comment by seanshpark** [comment]
x86 and RPi2 both works fine. OK to push.

File: tools/js2c.py [hunk 1]
```diff
@@ -93,32 +96,88 @@ def removeWhitespaces(code):
...
-NATIVE_STRUCT = '''
-struct native_mod {
-  const char* name;
-  const char* source;
-  const int length;
-};
-
-__attribute__ ((used)) static struct native_mod natives[] = {
-'''
+        NATIVE_STRUCT = '''
```
Comment (wateret) [inline tools/js2c.py]: I think there should be no indentation here.

File: tools/js2c.py [hunk 2]
```diff
@@ -93,32 +96,88 @@ def removeWhitespaces(code):
...
+        fmodule = open(path, 'r')
+        module = fmodule.read()
+        fmodule.close()
+
+        fmodule_wrapped = open(path + '.wrapped', 'w')
+        # FIXME
+        if name != 'iotjs':
+            fmodule_wrapped.write ("(function (a, b, c) {\n")
+            fmodule_wrapped.write (
+                "function wwwwrap(exports, require, module) {\n");
```
Comment (ILyoan) [inline tools/js2c.py]: Double wrapping not needed any longer.

File: tools/build.py [hunk 3]
```diff
@@ -611,7 +613,10 @@ def build_libhttpparser(option):
 def build_iotjs(option):
     # Run js2c
     os.chdir(SCRIPT_PATH)
-    check_run_cmd('python', ['js2c.py', option.buildtype, jerry_output_path])
+    check_run_cmd('python', ['js2c.py', option.buildtype,
+                             ('no_snapshot'
+                              if option.no_snapshot else 'snapshot'),
+                              jerry_output_path])
```
Comment (ILyoan) [inline tools/build.py]: Indent is not aligned.

File: src/iotjs_module_process.cpp [hunk 4]
```diff
@@ -156,15 +156,45 @@ JHANDLER_FUNCTION(Compile){
 
 
 JHANDLER_FUNCTION(CompileNativePtr){
-  JHANDLER_CHECK(handler.GetArgLength() == 1);
-  JHANDLER_CHECK(handler.GetArg(0)->IsObject());
-
-  JResult jres = WrapEval((const char*)handler.GetArg(0)->GetNative());
+  IOTJS_ASSERT(handler.GetArgLength() == 1);
+  IOTJS_ASSERT(handler.GetArg(0)->IsString());
```
Comment (ILyoan) [inline src/iotjs_module_process.cpp]: Using `JHANDLER_CHECK` here is recommended.

File: tools/build.py [hunk 5]
```diff
@@ -625,6 +630,11 @@ def build_iotjs(option):
     cmake_opt.append('-DTARGET_OS=' + option.target_os)
     cmake_opt.append('-DPLATFORM_DESCRIPT=' + platform_descriptor)
 
+    if not option.no_snapshot:
+        cmake_opt.append('-DSNAPSHOT=YES')
+    else:
+        cmake_opt.append('-DSNAPSHOT=NO')
+
```
Comment (ILyoan) [inline tools/build.py]: Can we just manipulate `option.compile_flag` directly here like;
```
if not option.no_snapshot:
  option.compile_flag.append('-DENABLE_SHAPSHOT`)
```

**Comment by chunseoklee** [comment]
@ILyoan enhanced as mentioned.

**Comment by chunseoklee** [comment]
@wateret updated as commented.

**Comment by wateret** [comment]
LGTM

**Comment by ILyoan** [comment]
LGTM

**Pull request merged**